rmcrackan/Libation
# Add validate and release workflows

### Description
This pull requests adds two workflows to the project.

* Validate will run on any commit to master or on any pull request targeting master. It will build all 4 versions of the code and upload them as artifacts of the workflow. These can be later downloaded for troubleshooting/verifying they run as expected. Pull requests will also be blocked from merging until the workflow completes without any errors.

* Release will run on any tag that starts with a `v`. This workflow builds the projects using the same shared steps as validate then creates a draft release, renames the artifacts to be in the correct format, and uploads them to the release. All you should need to do to make the release official is to edit the body of the release and remove the draft status.

### Testing
Testing for this will need to be done on my fork which has the pipelines already. I've added you as a collaborator.
* Make sure that the build workflow for a pull request runs successfully. You can create a test pull request and optionally commit something to it.
  * Check that the artifacts generated under the actions tab run correctly on their respective OSes. 
* Try creating a test tag that starts with a `v` 
  * It should run the release workflow, you can verify this in actions tab for the repo.
  * Verify that the draft release is created and that you are happy with the naming of everything.

### Other Changes
I made a couple other changes to support the workflows.
* I updated all build targets to use folders that match the capitalization of the OSes used in the folder structure, `macos` -> `MacOS`. This was to simplify the shared build workflow so that I only needed one variable for the OS.
* The publish profiles were all still on net6.0 even though the projects were on 7. I updated them all to 7 so that it would build correctly.

### Not Done
There are a couple of things that I left undone that could be added.
* It's not running any unit tests, this should be done, minimally, as part of the validate step. It probably makes more sense to just add it to the shared build workflow. If you share the commands you're using to run the unit tests I'll add them as part of this request.
* It should be possible to add a body to the release automatically, probably based on the commit messages since the last release. I haven't looked into this and it didn't seem that important compared to the quality of life enhancement that getting the workflows in place will give.
